CRM/Price/BAO/LineItem.php

```diff
@@ -819,60 +825,107 @@ protected function _getFinancialItemsToRecord($entityID, $entityTable, $contribu
-          $updateFinancialItemInfoValues['tax']['description'] = $taxTerm;
-          if ($lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['financial_type_id']) {
-            $updateFinancialItemInfoValues['tax']['financial_account_id'] = CRM_Contribute_BAO_Contribution::getFinancialAccountId($lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['financial_type_id']);
-          }
+      // if fee is changed to higher amount then update the existing financial item, this condition will hit ONLY when no payment has been made yet.
+      elseif (!empty($lineItemsToUpdate) &&
+        in_array($updateFinancialItemInfoValues['price_field_value_id'], $submittedPriceFieldValueIDs) &&
+        $updateFinancialItemInfoValues['entity_id'] == $lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['id'] &&
+        $updateFinancialItemInfoValues['amount'] > 0 && //escape if the financial item represent cancelling line-item
+        $lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['line_total'] != $updateFinancialItemInfoValues['amount']
+      ) {
+        $updateFinancialItemInfoValues['id'] = $previousFinancialItemID;
+        $updateFinancialItemInfoValues['amount'] = $lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['line_total'];
+        $updatedQty = $lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['qty'];
+        $updateFinancialItemInfoValues['description'] = ($updatedQty > 1 ? $updatedQty . ' of ' : '')  . $lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['label'];
+        if (!empty($lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]['tax_amount'])) {
+          $updateFinancialItemInfoValues['tax'] = $this->_getSalesTaxFinancialItem($lineItemsToUpdate[$updateFinancialItemInfoValues['price_field_value_id']]);
         }
-        $financialItemsArray[] = $updateFinancialItemInfoValues;
+
+        $financialItemsArray[$updateFinancialItemInfoValues['entity_id']] = $updateFinancialItemInfoValues;
       }
     }
 
     return $financialItemsArray;
   }
 
   /**
+   * Helper function to return sum of financial item's amount related to a line-item
+   * @param array $lineItemID
+   *
+   * @return float $financialItem
+   */
+  protected function _checkFinancialItemTotalAmountByLineItemID($lineItemID) {
+    return CRM_Core_DAO::singleValueQuery("
+      SELECT SUM(amount)
+      FROM civicrm_financial_item
+      WHERE entity_table = 'civicrm_line_item' AND entity_id = {$lineItemID}
+    ");
+  }
+
+  /**
+   * Helper function to return financial item related to Sales Tax
+   * @param array $lineItemInfo
+   *
+   * @return array $financialItem
+   */
+  protected function _getSalesTaxFinancialItem($lineItemInfo) {
+    static $taxTerm;
```

Comment:
further thought on this
1) there is enough caching in Civi::settings()->get IMHO
2) I think a function getInvoiceSetting() would be handy - these lines are repeated a few times - so it would look lik

```
function getInvoiceSetting($name) {
  return CRM_Utils_Array::value($name, Civi::settings()->get('contribution_invoice_settings'));
}
```

Doesn't really save any rows but might be more readable that the multiple places with 2 row retrievals at the moment